JENKINS-24207: Add new selector for Perforce/P4 that allows to update JIRA issues based on jobs

J-cztery opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jenkinsci/jira-plugin/95)
<!-- Reviewable:end -->

warden commented:
Nice! Can I ask you to create JIRA ticket for this in issues.jenkins-ci.org and reference that JIRA number in the commit message? That helps in the release notes :)

warden commented on src/main/java/hudson/plugins/jira/selector/p4/JobIssueSelector.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public Set<String> findIssueIds(Run<?, ?> run, JiraSite site, TaskListener listener) {
+        // TODO Auto-generated method stub
```
If this is not implemented yet, do we need to have it?

J-cztery replied:
No, definitely not. I just wanted to have your opinion on the direction. It got late last night so i had to stop. I will either implement p4 plugin or remove dependencies on it and the selector.

warden replied:
Great, sounds good, thanks!

warden commented on src/main/java/hudson/plugins/jira/selector/perforce/JobIssueSelector.java:
```diff
@@ -0,0 +1,152 @@
+       }else{
+           listener.getLogger().print("JIRA: The following JIRA issues will be updated: ");
+           Iterator<String> it = ids.iterator();
+           while(it.hasNext()){
```
You could use StringUtils.join()

J-cztery commented:
There already was a JIRA issue. Let me know if you have any other comments. I will squash the commits once it is final.

warden commented:
There's one thing I don't like - the static methods in DefaultIssueSelector and using them in the JobIssueSelector in that way.. Maybe you could move them AbstractIssueSelector and make non-static?

warden commented on src/main/java/hudson/plugins/jira/selector/jobs/JobIssueSelector.java:
```diff
@@ -0,0 +1,62 @@
+import hudson.plugins.jira.RunScmChangeExtractor;
+import hudson.plugins.jira.selector.AbstractIssueSelector;
+import hudson.plugins.jira.selector.DefaultIssueSelector;
+
```
Please also add JavaDoc describing this class

J-cztery commented:
@warden AbstractIssueSelector is not a good place for this methods. But I see what you are coming from. Please have a look at the changes i made (I think they are what you wanted but they are in DefaultIssueSelector rather than in AbstractIssueSelector) and see if you like them or not. 
The idea is that DefaultIssueSelector contains basic logic, i.e. take issues from parameters, carried over, dependent builds and from current builds. Any piece of this behaviour can be overridden. 

Added javadoc. 

Added config.jelly resources for selectors, including DefaultIssueSelector.

Strings are now taken from properties..

warden commented on src/main/java/hudson/plugins/jira/selector/DefaultIssueSelector.java:
```diff
@@ -27,64 +29,46 @@
-     *            
-     */
-    protected static void findIssues(Run<?, ?> build, Set<String> ids, Pattern pattern, TaskListener listener) {
+    protected static void findIssues(Run<?, ?> build, Set<String> issueIds, Pattern pattern, TaskListener listener) {
```
this static should not be needed here anymore :)

J-cztery replied:
If i remove it, i will break binary compatibility... 
Also, tests are using it... Tests can be reworked to not use this. But binary compatibility might be a bigger issue.

warden replied:
True, that's a valid point, lets keep it as is then for now.

warden commented on src/main/java/hudson/plugins/jira/selector/jobs/JobIssueSelector.java:
```diff
@@ -0,0 +1,34 @@
+package hudson.plugins.jira.selector.jobs;
```
Great changes, I like this now!
One thing only - as the "Job"s term are coming from Perforce and in Jenkins there is a similar term for a Job, I find this package name confusing. I propose to rename: hudson.plugins.jira.selector.jobs -> udson.plugins.jira.selector.perforce and move all classes there.
If you agree and this is done, I think we'll be good to merge!

J-cztery replied:
Ok. Makes sense. I though there will be other job based SCMs so it would make sense but i can't find any. So let's apply YAGNI principle here.

warden commented on src/main/java/hudson/plugins/jira/selector/jobs/perforce/PerforceJobIssueSelector.java:
```diff
@@ -0,0 +1,79 @@
+public class PerforceJobIssueSelector extends JobIssueSelector {
+    private static final Logger LOGGER = Logger.getLogger(PerforceJobIssueSelector.class.getName());
+
+    @Extension(optional = true)
```
not sure, but I think that @Extension on class level should be enough here

J-cztery replied:
Well, i copied what DefaultIssueSelector did.

warden commented on src/main/java/hudson/plugins/jira/selector/perforce/PerforceJobIssueSelector.java:
```diff
@@ -0,0 +1,78 @@
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+
+@Extension(optional = true)
```
Double annotation, one is already at DescriptorImpl.

J-cztery replied:
I am not sure what you are referring to. These are two different classes. Each of them is marked as an extension. Feel free to modify that code if you do not like it. 
I do not have any more time to test if @Extension annotates nested classes automatically and documenation does not state it clearly.

warden replied:
Per documentation:
```
 * Marks a field, a method, or a class for automatic discovery, so that Hudson can locate
 * implementations of {@link ExtensionPoint}s automatically.
 *
 * <p>
 * (In contrast, in earlier Hudson, the registration was manual.)
 *
 * <p>
 * In a simplest case, put this on your class, and Hudson will create an instance of it
 * and register it to the appropriate {@link ExtensionList}.
 *
 * <p>
 * If you'd like Hudson to call
 * a factory method instead of a constructor, put this annotation on your static factory
 * method. Hudson will invoke it and if the method returns a non-null instance, it'll be
 * registered. The return type of the method is used to determine which {@link ExtensionList}
 * will get the instance.
 *
 * Finally, you can put this annotation on a static field if the field contains a reference
 * to an instance that you'd like to register.
 *
```

See DefaultIssueSelector where @Extension is only at DescriptorImpl level.

warden commented on src/main/java/hudson/plugins/jira/selector/perforce/P4JobIssueSelector.java:
```diff
@@ -0,0 +1,74 @@
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+
+@Extension(optional = true)
```
same here

warden commented:
So I guess after removing the doubled annotations I'm okay with this one 👍

J-cztery commented:
@warden this is how much i can afford to contribute my work to this change. At this time it is either take it or leave it. I am sure there are others willing to contribute the changes you require. 
Thank you for your useful comments and proposals!
